### PR TITLE
Python bmi grid

### PR DESCRIPTION
**Pull request by hellkite500**
Add a gridded input and output variable to the python BMI unit test model.

## Additions

- A bmi_grid.py helper module
- Gridded input and output variable in the model

## Changes

- Fixed some string formatting and import issues in the stand alone test runner.

## Testing

1. Ran the stand alone test runner and validated the expected output print throughout the update loop.

## Notes

- Introduced some _conventions_ for dealing with setting and managing grid extents

## Todos

- A few todo/fixmes still in the BMI implementation, mostly related to cleaning up some hard coded magic numbers and possibly refactoring the relationship between grids and variables to formalize their connection through some helper utilities.

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [x] Linux
- [x] MacOS

### CODE REVIEW

**Comment by hellkite500** [comment]
I have identified and "fixed" all the breaking tests in `Bmi_Py_Adapter_Test.cpp`, but the `Bmi_Py_Formulation_Test` failures are more challenging, since we don't yet have a gridded data provider...

**Review by aaraney — COMMENTED** [review]
Cool stuff! I just left some stylistic comments as I grazed through the code to better understand what you're thinking.

File: extern/test_bmi_py/bmi_grid.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    unstructured = "unstructured",
+    structured_quadrilateral = "structured_quadrilaterl",
+    rectilinear = "rectilinear",
+    uniform_rectilinear = "uniform_rectilinear"
+
+class Grid():
+    """
+        Structure for holding required BMI meta data for any grid intended to be used via BMI
+    """
+    def __init__(self, id: int, rank: int , type: GridType):
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/bmi_grid.py:33]: linting tools? What are those 🤣

File: extern/test_bmi_py/bmi_grid.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        Args:
+            id (int): User defined identifier for this grid
+            rank (int): The number of dimensions of the grid
+            type (GridType): The type of BMI grid this meta data represents
+        """
+        self._id: int = id
+        self._rank: int = rank
+        self._size: int = 0 #FIXME make property
+        self._type: GridType = type #FIXME validate type/rank?
+        self._shape: Tuple[int] = None #tuple of size rank
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/bmi_grid.py:45]: Yeah, these are all going to change to `np.array` though...for reasons of accessing cleanly from ngen BMI adapter...

File: extern/test_bmi_py/bmi_model.py [hunk 3]
```diff
@@ -27,28 +27,40 @@ def __init__(self):
         self._model = None
         self.var_array_lengths = 1
 
+        self.grid_0: Grid = Grid(0, 0, GridType.scalar) #Grid 0 is a 0 dimension "grid" for scalars
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/bmi_model.py:30]: I'm always torn between the readability of kwargs and the notion that they are "optional"...positional, required args, vs required "optional" kwargs...

File: extern/test_bmi_py/bmi_model.py [hunk 4]
```diff
@@ -200,7 +216,14 @@ def get_value(self, var_name: str, dest: np.ndarray) -> np.ndarray:
...
         """
-        dest[:] = self.get_value_ptr(var_name)
+        if var_name == "grid:count":
+            dest[...] = 2
+        elif var_name == "grid:ids":
+            dest[:] = [self.grid_0.id, self.grid_1.id]
+        elif var_name == "grid:ranks":
+            dest[:] = [self.grid_0.rank, self.grid_1.rank]
+        else:
+            dest = self.get_value_ptr(var_name)
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/bmi_model.py:226]: hmm...probably

File: extern/test_bmi_py/run_bmi_model.py [hunk 5]
```diff
@@ -1,10 +1,32 @@
...
+    """_summary_
 
+    Args:
+        model (Bmi): the Bmi model to query
+        name (str): the name of the variable to query
+
+    Returns:
+        ndarray: numpy array with the value of the variable marshalled through BMI
+    """
+    #TODO most (if not all) of this can be cached...unless the grid can change?
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/run_bmi_model.py:22]: It wouldn't be too difficult to differentiate a "static" vs a "dynamic" grid with some additional meta data...

**Comment by hellkite500** [comment]
The last 4 commits get all the existing Adapter unit tests passing on both macos and ubuntu.  Will add some new tests for more grid definitions and operations.  Still haven't completely addressed the Formulation tests, but some discussion around using sloth to mock up grid data seems like it might work for now.

File: extern/test_bmi_py/bmi_grid.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aaraney) [reply extern/test_bmi_py/bmi_grid.py:33]: 😆

File: extern/test_bmi_py/bmi_grid.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaraney) [reply extern/test_bmi_py/bmi_grid.py]: leave a TODO?

File: extern/test_bmi_py/bmi_grid.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+if TYPE_CHECKING:
+    from typing import Tuple
+    from numpy.typing import NDArray
+
+class GridType(str, Enum):
+    """
+        Enumeration of supported BMI grid types (https://bmi.readthedocs.io/en/stable/#get-grid-type)
+    """
+    scalar = "scalar",
```
Comment (mattw-nws) [inline extern/test_bmi_py/bmi_grid.py:21]: Enum standard seems to be UPPERCASE but I can't find a reference... PEP 8 doesn't specifically address Enums?

File: extern/test_bmi_py/bmi_grid.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    def shape(self) -> 'NDArray[np.int32]':
+        """The shape of the grid (the size of each dimension)
+
+        Returns:
+            Tuple[int]: size of each dimension
+        """
+        return self._shape
+    
+    @shape.setter
+    def shape(self, shape: 'NDArray[np.int32]') -> None:
```
Comment (mattw-nws) [inline extern/test_bmi_py/bmi_grid.py:114]: Is this type hint for the input wrong?

File: extern/test_bmi_py/bmi_grid.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/bmi_grid.py]: Fixed in 158fa584ca89dbfcae8f3816343996b20132aac4

File: extern/test_bmi_py/bmi_grid.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    def type(self) -> GridType:
+        """The type of BMI grid.
+
+        Returns:
+            GridType: bmi grid type
+        """
+        return self._type
+    
+    @property
+    def shape(self) -> 'Tuple[int]':
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/bmi_grid.py]: fixed in 158fa584ca89dbfcae8f3816343996b20132aac4

File: extern/test_bmi_py/bmi_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hellkite500) [reply extern/test_bmi_py/bmi_model.py]: fixed in ac64f26a9255e9cce0d8c9d9d325e401cf7130ee

File: extern/test_bmi_py/bmi_grid.py (hunk 7 above)
Comment (hellkite500) [reply extern/test_bmi_py/bmi_grid.py:189]: I'm pretty sure this is resolved in the latest changes, but let me know if I'm missing something.

**Comment by hellkite500** [comment]
Still have `GetGridX/Y/Z` to complete, then all the rectilinear grid types should be fully supported.

File: extern/test_bmi_py/bmi_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattw-nws) [reply extern/test_bmi_py/bmi_model.py:33]: yeah we need `fn(required_arg!=0, optional_arg=1)`

File: extern/test_bmi_py/run_bmi_model.py [hunk 9]
```diff
@@ -1,10 +1,32 @@
 from pathlib import Path
 
 import numpy as np
-
+np.set_printoptions(precision=2)
 # This is the BMI LSTM that we will be running
```
Comment (mattw-nws) [inline extern/test_bmi_py/run_bmi_model.py:5]: Not to be pedantic, but...

Comment (hellkite500) [reply extern/test_bmi_py/run_bmi_model.py:5]: ???

Comment (mattw-nws) [reply extern/test_bmi_py/run_bmi_model.py:5]: oh that's on the wrong line somehow. Line 5... it's not LSTM 😝

Comment (hellkite500) [reply extern/test_bmi_py/run_bmi_model.py]: ahh....happy now?

File: test/realizations/catchments/Bmi_Py_Adapter_Test.cpp [hunk 10]
```diff
@@ -106,17 +106,19 @@ class Bmi_Py_Adapter_Test : public ::testing::Test {
...
-    std::vector<std::string> expected_input_var_types = { "float64", "float64" };
-    std::vector<int> expected_input_var_item_sizes = { 8, 8 };
-    std::vector<std::string> expected_output_var_types = { "float64", "float64", "float64" };
-    std::vector<int> expected_output_var_item_sizes = { 8, 8, 8 };
-    int expected_grid_rank = 1;
-    int expected_grid_size = 1;
+    std::vector<std::string> expected_output_var_names = { "OUTPUT_VAR_1", "OUTPUT_VAR_2", "OUTPUT_VAR_3", "GRID_VAR_2", "GRID_VAR_3" };
+    std::vector<std::string> expected_input_var_names = { "INPUT_VAR_1", "INPUT_VAR_2", "GRID_VAR_1"};
+    std::vector<std::string> expected_output_var_locations = { "node", "node", "node", "node" };
+    std::vector<int> expected_output_var_grids = { 0, 0, 0, 1 };
```
Comment (hellkite500) [inline test/realizations/catchments/Bmi_Py_Adapter_Test.cpp:112]: Should probably update this to consider grid 2 (the 3D grid for GRID_VAR_3)

Comment (hellkite500) [reply test/realizations/catchments/Bmi_Py_Adapter_Test.cpp:112]: as well as the rest of the meta data functions for GRID_VAR_3

**Review by donaldwj — APPROVED** [review]
We will need to look closely at the sections that where already marked as fix for only working if they where in layer one position but that can be handled when layer are merged.

**Comment by mattw-nws** [comment]
Failing test fixed in master--merging

**Pull request merged**